client/allocrunner/taskrunner/template/template_test.go

```diff
@@ -1914,3 +1915,244 @@ WAIT_LOOP:
+				},
+			},
+		},
+		{
+			"template-override",
+			&config.ClientTemplateConfig{
+				MaxStale:           helper.TimeToPtr(5 * time.Second),
+				BlockQueryWaitTime: helper.TimeToPtr(60 * time.Second),
+				Wait:               waitConfig.Copy(),
+				ConsulRetry:        retryConfig.Copy(),
+				VaultRetry:         retryConfig.Copy(),
+			},
+			&TaskTemplateManagerConfig{
+				ClientConfig: clientConfig,
+				VaultToken:   "token",
+				EnvBuilder:   taskenv.NewBuilder(clientConfig.Node, allocWithOverride, allocWithOverride.Job.TaskGroups[0].Tasks[0], clientConfig.Region),
+			},
+			&config.Config{
+				TemplateConfig: &config.ClientTemplateConfig{
+					MaxStale:           helper.TimeToPtr(5 * time.Second),
+					BlockQueryWaitTime: helper.TimeToPtr(60 * time.Second),
+					Wait:               waitConfig.Copy(),
+					ConsulRetry:        retryConfig.Copy(),
+					VaultRetry:         retryConfig.Copy(),
+				},
+			},
+			&templateconfig.TemplateConfig{
+				Wait: &templateconfig.WaitConfig{
+					Enabled: helper.BoolToPtr(true),
+					Min:     helper.TimeToPtr(2 * time.Second),
+					Max:     helper.TimeToPtr(12 * time.Second),
+				},
+			},
+		},
+		{
+			"bounds-override",
+			&config.ClientTemplateConfig{
+				MaxStale:           helper.TimeToPtr(5 * time.Second),
+				BlockQueryWaitTime: helper.TimeToPtr(60 * time.Second),
+				Wait:               waitConfig.Copy(),
+				WaitBounds: &config.WaitConfig{
+					Min: helper.TimeToPtr(3 * time.Second),
+					Max: helper.TimeToPtr(11 * time.Second),
+				},
+				ConsulRetry: retryConfig.Copy(),
+				VaultRetry:  retryConfig.Copy(),
+			},
+			&TaskTemplateManagerConfig{
+				ClientConfig: clientConfig,
+				VaultToken:   "token",
+				EnvBuilder:   taskenv.NewBuilder(clientConfig.Node, allocWithOverride, allocWithOverride.Job.TaskGroups[0].Tasks[0], clientConfig.Region),
+				Templates: []*structs.Template{
+					{
+						Wait: &structs.WaitConfig{
+							Min: helper.TimeToPtr(2 * time.Second),
+							Max: helper.TimeToPtr(12 * time.Second),
+						},
+					},
+				},
+			},
+			&config.Config{
+				TemplateConfig: &config.ClientTemplateConfig{
+					MaxStale:           helper.TimeToPtr(5 * time.Second),
+					BlockQueryWaitTime: helper.TimeToPtr(60 * time.Second),
+					Wait:               waitConfig.Copy(),
+					WaitBounds: &config.WaitConfig{
+						Min: helper.TimeToPtr(3 * time.Second),
+						Max: helper.TimeToPtr(11 * time.Second),
+					},
+					ConsulRetry: retryConfig.Copy(),
+					VaultRetry:  retryConfig.Copy(),
+				},
+			},
+			&templateconfig.TemplateConfig{
+				Wait: &templateconfig.WaitConfig{
+					Enabled: helper.BoolToPtr(true),
+					Min:     helper.TimeToPtr(3 * time.Second),
+					Max:     helper.TimeToPtr(11 * time.Second),
+				},
+			},
+		},
+	}
+
+	for _, _case := range cases {
+		t.Run(_case.Name, func(t *testing.T) {
+			// monkey patch the client config with the version of the ClientTemplateConfig we want to test.
```

Comment:
This works, but for future reference for tests like this where there's a lot of nested objects DRY-ing it up works against you a bit. It'd probably have been fine to have one big input `TaskTemplateManagerConfig` and one big "expected" `TemplateConfig` that you can compare with a single `require.Equal`.